datasets/pl/finanse_sanctions/crawler.py

```diff
@@ -16,88 +16,107 @@
-    entity.add("address", row.pop("location full") or None)
-    entity.add("country", row.pop("location country") or None)
+    h.apply_date(entity, "birthDate", row.pop("data_urodzenia"))
 
-    entity.add("nationality", row.pop("narodowość"))
-    entity.add("topics", "sanction")
+    other_information = row.pop("inne_informacje")
+    res = context.lookup("other_information", other_information)
+    if not res:
+        context.log.warning("No extra information lookup found", row=row)
+    else:
+        entity.add(
+            "nationality",
+            res.properties.get("nationality"),
+            lang="pol",
+            original_value=other_information,
+        )
+        entity.add(
+            "country",
+            res.properties.get("country"),
+            lang="pol",
+            original_value=other_information,
+        )
+        entity.add(
+            "address",
+            res.properties.get("address"),
+            lang="pol",
+            original_value=other_information,
+        )
 
+    entity.add("topics", "sanction")
     sanction = h.make_sanction(context, entity)
-    h.apply_date(sanction, "listingDate", row.pop("Data umieszczenia na liście"))
+    h.apply_date(sanction, "listingDate", row.pop("data_umieszczenia_na_liscie"))
     sanction.add(
-        "reason", collapse_spaces(row.pop("Uzasadnienie wpisu na listę")), lang="pol"
+        "reason",
+        collapse_spaces(row.pop("uzasadnienie_wpisu_na_liste_url")),
+        lang="pol",
     )
     sanction.add("program", POLAND_PROGRAM, "pol")
 
     context.emit(entity, target=True)
     context.emit(sanction)
 
-    context.audit_data(row)
+    context.audit_data(
+        row,
+        ignore=[
+            "lp",  # ID, we don't trust this to be stable
+            "uzasadnienie_wpisu_na_liste",  # Reason link text, we only care about the url
+        ],
+    )
 
 
-def check_updates(context: Context):
+def crawl(context: Context):
     doc = context.fetch_html(context.dataset.url)
     doc.make_links_absolute(context.dataset.url)
-    materials = doc.findall(".//a[@class='file-download']")
 
-    # Process the materials
-    if len(materials) == 0:
-        context.log.warning("No materials downloads found")
-    else:
-        for material in materials:
-            url = material.get("href")
-            if url in KNOWN_HASHES:
-                h.assert_url_hash(context, url, KNOWN_HASHES[url])
-            else:
-                context.log.warning(
-                    "Unknown materials download URL. Check if we want to scrape it.",
-                    url=url,
-                )
+    xlsx_link_element = doc.xpath(
+        ".//a[@class='file-download' and contains(text(), 'Lista osób i podmiotów, wobec których stosuje się szczególne środki ograniczające na podstawie art. 118 ust. 1 pkt 2 (wersja xlsx)')]"
+    )
+    if len(xlsx_link_element) != 1:
+        context.log.error("Could not find XLSX link element")
+        return
```

Comment:
A lot of the time it's best for a crawler to exit and don't publish when they're missing something important. E.g. in this case we don't want to publish a new dataset without the entities in this file, just because they changed the link text. It's better for it to exit loudly so we can come and fix it.

The way this is currently implemented, it should behave that way. If we don't find the file we exit before we export the UN stubs, so the dataset will be empty and won't publish. But it's worth bearing this requirement in mind.

There are exceptions - really big datasets where we sometimes struggle to do complete crawls sometimes allow for some failure, but we try to limit that.

Reply:
Thanks for the explanation! I'm not quite sure if you were happy with what I did, I just changed it to raise a `RuntimeError`, maybe that's even better?

Reply:
what you did was fine because it would do the right thing. Raising is definitely my personal preference because it's robust to future changes in the crawler, but I'm not sure if we have a general preference.